### PR TITLE
[Wallet] move "load wallet phase" to CWallet

### PR DESCRIPTION
**Pull request by jonasschnelli**
Another simple and easy-to-review wallet/init.cpp refactoring

### CODE REVIEW

**Comment by kirkalx** [comment]
Looks good to me... utACK.

**Comment by kirkalx** [comment]
Does this need a bump to get travis to check it @jonasschnelli ?

**Comment by jonasschnelli** [comment]
Force pushed (no changes), this should give Travis a kick.

**Comment by laanwj** [comment]
Concept ACK, this is a nice step toward factoring out the wallet code.
 
I do think the return condition is a bit messy, e.g. using both a NULL return value as well as a non-empty errorString as an error condition:
```patch
 +        if (!pwalletMain)
 +            return false;
...
 +        if (!errorString.empty())
 +            return InitError(errorString);
```
Ideally, InitLoadWallet would always return NULL if an error happened, and always fill errorString.
I understand that this is a change with more impact, though: e.g. make `walletInstance` a scoped pointer, which is released only on success.

**Comment by kirkalx** [comment]
From the initial work I've done on updating some of Jonas's other wallet patches, this is familiar and I agree it can be confusing, especially in some of the more complicated changes where wallet code calls some other wallet code with its own error handling. The intention is to end up with at least one (and preferably only one) error message in the log if something goes wrong. Worst case scenario if we screw up is that the program just shuts down without an error being reported.

**Comment by jonasschnelli** [comment]
Agree with @laanwj. I wanted to keep the changes at a minimum because this is not the first try to get this into master.

Will adapt a better return code handling.

**Comment by kirkalx** [comment]
Sweet, utACK 32ffefc12ba5a5956d64a7c207f0070a11041339. Sorry, no dev machine at present so can't test...

**Comment by laanwj** [comment]
> Agree with @laanwj. I wanted to keep the changes at a minimum because this is not the first try to get this into master.

Yes, changing this may be too risky right now, at least this is straightforward and easy to review.

**Comment by laanwj** [comment]
Needs rebase after #7576

**Comment by jonasschnelli** [comment]
Rebased.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2991,6 +2991,165 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
...
+    uiInterface.InitMessage(_("Loading wallet..."));
+
+    int64_t nStart = GetTimeMillis();
+    bool fFirstRun = true;
+    CWallet *walletInstance = new CWallet(strWalletFile);
+    DBErrors nLoadWalletRet = walletInstance->LoadWallet(fFirstRun);
+    if (nLoadWalletRet != DB_LOAD_OK)
+    {
+        if (nLoadWalletRet == DB_CORRUPT)
+            errorString += _("Error loading wallet.dat: Wallet corrupted") + "\n";
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Nit: Please don't mix concatenations (`+=`)and assignments (`=`). You refactored this such that there can only be one error in the `errorString`, so an assignment would be sufficient.

Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Though,  if you prefer to keep the old style concatenation,  please make all assignments to be concatenations.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2991,6 +2991,165 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
...
+            {
+                errorString += _("Cannot write default address") += "\n";
+                return NULL;
+            }
+        }
+
+        walletInstance->SetBestChain(chainActive.GetLocator());
+    }
+
+    LogPrintf("%s", errorString);
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: You'd have to move this above the `return NULL`.

**Comment by MarcoFalke** [comment]
Concept ACK  4841d7d

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: My main focus was it to make the change as "moveonly" as possible. The current code also concats error strings (https://github.com/bitcoin/bitcoin/pull/7577/files#diff-c865a8939105e6350a50af02766291b7L1454).

IMO best strategy is to move the wallet code away from init.cpp, then optimize it. This changes was already included in two closed PR (which where closed due to inactivity and lack of review).
We probably should distinct between refactor/"moveonly-ish" PRs and behavior change PRs.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Right. This is dead code now. Move the log print to the init.cpp error handling (L1434).

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (laanwj) [inline src/wallet/wallet.cpp]: It is ugly but was already the case in the previous version. This literally moves the code. Let's leave improvements here to later pulls.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Thx for the quick fix. Good this is merged now.

**Pull request merged**